packages/datadog-plugin-next/src/index.js

```diff
@@ -0,0 +1,132 @@
+'use strict'
+
+function createWrapHandleRequest (tracer, config) {
+  return function wrapHandleRequest (handleRequest) {
+    return function handleRequestWithTrace (req, res, pathname, query) {
+      return trace(tracer, config, req, res, () => handleRequest.apply(this, arguments))
+    }
+  }
+}
+
+function createWrapHandleApiRequest (tracer, config) {
+  return function wrapHandleApiRequest (handleApiRequest) {
+    return function handleApiRequestWithTrace (req, res, pathname, query) {
+      return trace(tracer, config, req, res, () => {
+        const promise = handleApiRequest.apply(this, arguments)
+
+        promise.then(handled => {
+          if (!handled) return
+
+          const page = getPageFromPath(pathname, this.dynamicRoutes)
+
+          addPage(req, page)
+        })
+
+        return promise
+      })
+    }
+  }
+}
+
+function createWrapRenderToHTML (tracer, config) {
+  return function wrapRenderToHTML (renderToHTML) {
+    return function renderToHTMLWithTrace (req, res, pathname, query, parsedUrl) {
+      return trace(tracer, config, req, res, () => renderToHTML.apply(this, arguments))
+    }
+  }
+}
+
+function createWrapRenderToHTMLWithComponents (tracer, config) {
+  return function wrapHandleRenderToHTMLWithComponents (renderToHTMLWithComponents) {
+    return function renderToHTMLWithComponentsWithTrace (req, res, page) {
+      addPage(req, page)
+
+      return renderToHTMLWithComponents.apply(this, arguments)
+    }
+  }
+}
+
+function getPageFromPath (page, dynamicRoutes = []) {
+  for (const dynamicRoute of dynamicRoutes) {
+    if (dynamicRoute.page.startsWith('/api') && dynamicRoute.match(page)) {
+      return dynamicRoute.page
+    }
+  }
+
+  return page
+}
+
+function trace (tracer, config, req, res, handler) {
+  const scope = tracer.scope()
+
+  if (req._datadog_next) return scope.activate(req._datadog_next.span, handler)
+
+  const childOf = scope.active()
+  const tags = {
+    'service.name': config.service || `${tracer._service}-next`,
+    'resource.name': req.method,
+    'span.type': 'web',
+    'span.kind': 'server',
+    'http.status_code': res.statusCode,
+    'http.method': req.method
+  }
+  const span = tracer.startSpan('next.request', { childOf, tags })
+
+  req._datadog_next = { span }
+
+  const promise = scope.activate(span, handler)
+
+  promise.then(() => finish(span, config, req, res), err => {
+    span.setTag('error', err)
+    finish(span, config, req, res)
+  })
+
+  return promise
+}
+
+function addPage (req, page) {
+  req._datadog_next.span.addTags({
+    'resource.name': `${req.method} ${page}`,
```

Comment:
Do we always have the `page`? I.e. do we ever end up having `GET ` or something like that with a trailing space?

Reply:
I'm pretty sure that it's not possible since these methods expect to explicitly be provided a page, but there might be some default where it's not needed. I'll trim it.